File: automation-test-playwright/pages/components/TableComponent.ts
import { type Locator, type Page } from "@playwright/test";
import { WaitHelper } from "@helpers-browser/WaitHelper";

export class TableComponent {
  constructor(
    private readonly page: Page,
    private readonly bodySelector: string,
    private readonly emptySelector = ".empty-state"
  ) {}

  body(): Locator {
    return this.page.locator(this.bodySelector);
  }

  rows(): Locator {
    return this.body().locator("tr");
  }

  rowByText(text: string | RegExp): Locator {
    return this.rows().filter({ hasText: text }).filter({ visible: true }).first();
  }

  async waitForDataOrEmpty(): Promise<void> {
    await WaitHelper.until(async () => {
      const hasRows = (await this.rows().count()) > 0;
      const hasEmpty = await this.page
        .locator(this.emptySelector)
        .isVisible()
        .catch(() => false);
      return hasRows || hasEmpty;
    });
  }
}

File: automation-test-selenium/src/main/java/org/suntower/pages/components/TableComponent.java
package org.suntower.pages.components;

import java.util.List;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.suntower.core.BasePage;

public class TableComponent extends BasePage {
  private final By body;
  private final By emptyState;

  public TableComponent(WebDriver driver, String bodySelector) {
    this(driver, bodySelector, ".empty-state");
  }

  public TableComponent(WebDriver driver, String bodySelector, String emptySelector) {
    super(driver);
    this.body = css(bodySelector);
    this.emptyState = css(emptySelector);
  }

  public By body() {
    return body;
  }

  public By rowsLocator() {
    return By.cssSelector(selector(body) + " tr");
  }

  public List<WebElement> rows() {
    return driver.findElements(rowsLocator());
  }

  public WebElement rowByText(String text) {
    waitForCondition(
        () ->
            rows().stream()
                .anyMatch(
                    row -> {
                      try {
                        return row.isDisplayed() && row.getText().contains(text);
                      } catch (RuntimeException ignored) {
                        return false;
                      }
                    }),
        "No visible table row contained text: " + text);
    return rows().stream()
        .filter(
            row -> {
              try {
                return row.isDisplayed() && row.getText().contains(text);
              } catch (RuntimeException ignored) {
                return false;
              }
            })
        .findFirst()
        .orElseThrow();
  }

  public void waitForDataOrEmpty() {
    waitForCondition(
        () -> !rows().isEmpty() || isVisible(emptyState),
        "Table did not show data rows or empty state.");
  }

  private String selector(By locator) {
    String raw = locator.toString();
    if (!raw.startsWith("By.cssSelector: ")) {
      throw new IllegalArgumentException("TableComponent requires CSS selectors.");
    }
    return raw.replace("By.cssSelector: ", "");
  }
}
